### Import required dependencies

In [1]:
import pandas as pd
import os

## Deliverable 1: Collect the Data

To collect the data that you’ll need, complete the following steps:

1. Using the Pandas `read_csv` function and the `os` module, import the data from the `new_full_student_data.csv` file, and create a DataFrame called student_df. 

2. Use the head function to confirm that Pandas properly imported the data.


In [2]:
# Create the path and import the data
full_student_data = os.path.join('../Resources/new_full_student_data.csv')
student_df = pd.read_csv(full_student_data)

In [3]:
# Verify that the data was properly imported
student_df.head()

student_id      student_name grade           school_name  reading_score  \
0   103880842     Travis Martin   9th  Sullivan High School           59.0   
1    45069750     Michael Brown   9th     Dixon High School           94.7   
2    45024902   Gabriela Lucero   9th    Wagner High School           89.0   
3    62582498  Susan Richardson   9th     Silva High School           69.7   
4    16437227      Sherry Davis  11th    Bowers High School            NaN   

   math_score school_type  school_budget  
0        88.2      Public         961125  
1        73.5     Charter         870334  
2        70.4      Public         846745  
3        80.3      Public         991918  
4        27.5      Public         848324

## Deliverable 2: Prepare the Data

To prepare and clean your data for analysis, complete the following steps:
    
1. Check for and remove all rows with `NaN`, or missing, values in the student DataFrame. 

2. Check for and remove all duplicate rows in the student DataFrame.

3. Use the `str.replace` function to remove the "th" from the grade levels in the grade column.

4. Check data types using the `dtypes` property.

5. Remove the "th" suffix from every value in the grade column using `str` and `replace`.

6. Change the grade colum to the `int` type and verify column types.

7. Use the head (and/or the tail) function to preview the DataFrame.

In [4]:
# Check for null values
student_df.isna().sum()

student_id          0
student_name        0
grade               0
school_name         0
reading_score    1968
math_score        982
school_type         0
school_budget       0
dtype: int64

In [5]:
# Drop rows with null values and verify removal
student_df = student_df.dropna()
student_df
student_df.isna().sum()

student_id       0
student_name     0
grade            0
school_name      0
reading_score    0
math_score       0
school_type      0
school_budget    0
dtype: int64

In [6]:
# Check for duplicated rows
student_df.duplicated()
student_df.duplicated().sum()

1836

In [7]:
# Drop duplicated rows and verify removal
student_df.drop_duplicates(inplace=True)
student_df.duplicated()

0        False
1        False
2        False
3        False
5        False
         ...  
19508    False
19509    False
19511    False
19512    False
19513    False
Length: 14831, dtype: bool

In [8]:
student_df

student_id      student_name grade             school_name  \
0       103880842     Travis Martin   9th    Sullivan High School   
1        45069750     Michael Brown   9th       Dixon High School   
2        45024902   Gabriela Lucero   9th      Wagner High School   
3        62582498  Susan Richardson   9th       Silva High School   
5        74579444   Cynthia Johnson   9th  Montgomery High School   
...           ...               ...   ...                     ...   
19508    83985333   Deborah Sanders  10th       Silva High School   
19509   109236636     Robert Sawyer  12th       Silva High School   
19511    95516554        Megan Gill  11th      Wagner High School   
19512    65050383        Lori Stone  11th      Bowers High School   
19513    34720657       Anna Jensen  12th  Montgomery High School   

       reading_score  math_score school_type  school_budget  
0               59.0        88.2      Public         961125  
1               94.7        73.5     Charter         870334  
2               89.0        70.4      Public         846745  
3               69.7        80.3      Public         991918  
5               63.5        76.9     Charter         893368  
...              ...         ...         ...            ...  
19508           60.5        64.6      Public         991918  
19509           43.3        27.2      Public         991918  
19511           93.9        84.1      Public         846745  
19512           94.6        70.9      Public         848324  
19513           82.3        42.4     Charter         893368  

[14831 rows x 8 columns]

In [9]:
# Check data types

student_df.dtypes



student_id         int64
student_name      object
grade             object
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

In [10]:
# Examine the grade column to understand why it is not an int
student_df["grade"]

0         9th
1         9th
2         9th
3         9th
5         9th
         ... 
19508    10th
19509    12th
19511    11th
19512    11th
19513    12th
Name: grade, Length: 14831, dtype: object

In [11]:
# Remove the non-numeric characters and verify the contents of the column
student_df['grade'] = student_df['grade'].str.replace('th', '')
student_df["grade"]

0         9
1         9
2         9
3         9
5         9
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 14831, dtype: object

In [12]:
# Change the grade column to the int type and verify column types
student_df['grade'] = student_df['grade'].astype(int)
student_df.dtypes

student_id         int64
student_name      object
grade              int64
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

## Deliverable 3: Summarize the Data

Describe the data using summary statistics on the data as a whole and on individual columns.

1. Generate the summary statistics for each DataFrame by using the `describe` function.

2. Display the mean math score using the `mean` function. 

2. Store the minimum reading score as `min_reading_score`.

In [13]:
# Display summary statistics for the DataFrame
student_df.describe()

student_id         grade  reading_score    math_score  school_budget
count  1.483100e+04  14831.000000   14831.000000  14831.000000   14831.000000
mean   6.975296e+07     10.355539      72.357865     64.675733  893742.749107
std    3.452909e+07      1.097728      15.224590     15.844093   53938.066467
min    1.000906e+07      9.000000      10.500000      3.700000  817615.000000
25%    3.984433e+07      9.000000      62.200000     54.500000  846745.000000
50%    6.965978e+07     10.000000      73.800000     65.300000  893368.000000
75%    9.927449e+07     11.000000      84.000000     76.000000  956438.000000
max    1.299997e+08     12.000000     100.000000    100.000000  991918.000000

In [14]:
# Display the mean math score using the mean function
student_df["math_score"].mean()

64.67573326141189

In [15]:
# Store the minimum reading score as min_reading_score
min_reading_score = student_df["reading_score"].min()
min_reading_score

10.5

## Deliverable 4: Drill Down into the Data

Drill down to specific rows, columns, and subsets of the data.

To drill down into the data, complete the following steps:

1. Use `loc` to display the grade column.

2. Use `iloc` to display the first 3 rows and columns 3, 4, and 5.

3. Show the rows for grade nine using `loc`.

4. Store the row with the minimum overall reading score as `min_reading_row` using `loc` and the `min_reading_score` found in Deliverable 3.

5. Find the reading scores for the school and grade from the output of step three using `loc` with multiple conditional statements.

6. Using conditional statements and `loc` or `iloc`, find the mean reading score for all students in grades 11 and 12 combined.

In [16]:
# Use loc to display the grade column
student_df.loc[:,"grade"]

0         9
1         9
2         9
3         9
5         9
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 14831, dtype: int64

In [17]:
# Use `iloc` to display the first 3 rows and columns 3, 4, and 5.
student_df.iloc[:, [3, 4, 5]]

school_name  reading_score  math_score
0        Sullivan High School           59.0        88.2
1           Dixon High School           94.7        73.5
2          Wagner High School           89.0        70.4
3           Silva High School           69.7        80.3
5      Montgomery High School           63.5        76.9
...                       ...            ...         ...
19508       Silva High School           60.5        64.6
19509       Silva High School           43.3        27.2
19511      Wagner High School           93.9        84.1
19512      Bowers High School           94.6        70.9
19513  Montgomery High School           82.3        42.4

[14831 rows x 3 columns]

###### Select the rows for grade nine and display their summary statistics using `loc` and `describe`.
grade_nine_df = student_df.loc[student_df["grade"] == 9]
grade_nine_df


In [18]:
grade_nine_df = student_df.loc[student_df["grade"] == 9] 
grade_nine_df


student_id      student_name  grade             school_name  \
0       103880842     Travis Martin      9    Sullivan High School   
1        45069750     Michael Brown      9       Dixon High School   
2        45024902   Gabriela Lucero      9      Wagner High School   
3        62582498  Susan Richardson      9       Silva High School   
5        74579444   Cynthia Johnson      9  Montgomery High School   
...           ...               ...    ...                     ...   
19477    14165038     William Smith      9       Dixon High School   
19478    46569330     Angie Flowers      9      Bowers High School   
19481    99423494     Sherry Rogers      9       Silva High School   
19493    68274108       Mary Murray      9      Wagner High School   
19497    56856952     Joshua Holmes      9      Wagner High School   

       reading_score  math_score school_type  school_budget  
0               59.0        88.2      Public         961125  
1               94.7        73.5     Charter         870334  
2               89.0        70.4      Public         846745  
3               69.7        80.3      Public         991918  
5               63.5        76.9     Charter         893368  
...              ...         ...         ...            ...  
19477           69.8        65.0     Charter         870334  
19478           50.4        61.7      Public         848324  
19481           47.0        88.4      Public         991918  
19493           63.4        53.4      Public         846745  
19497           77.2        76.5      Public         846745  

[4132 rows x 8 columns]

In [19]:
# Store the row with the minimum overall reading score as `min_reading_row`
# using `loc` and the `min_reading_score` found in Deliverable 3.
min_reading_score = student_df["reading_score"].min()
min_reading_row = student_df.loc[student_df["reading_score"] == min_reading_score]
min_reading_row

student_id    student_name  grade        school_name  reading_score  \
3706    81758630  Matthew Thomas     10  Dixon High School           10.5   

      math_score school_type  school_budget  
3706        58.4     Charter         870334

In [20]:
# Use loc with conditionals to select all reading scores from 10th graders at Dixon High School.
student_df.loc[(student_df["grade"] == 10) & (student_df["school_name"] == "Dixon High School") & (student_df["reading_score"])][["grade", "school_name", "reading_score"]]


grade        school_name  reading_score
45        10  Dixon High School           71.1
60        10  Dixon High School           59.5
69        10  Dixon High School           88.6
94        10  Dixon High School           81.5
100       10  Dixon High School           95.3
...      ...                ...            ...
19283     10  Dixon High School           52.9
19306     10  Dixon High School           58.0
19344     10  Dixon High School           38.0
19368     10  Dixon High School           84.4
19445     10  Dixon High School           43.9

[569 rows x 3 columns]

In [21]:
# Find the mean reading score for all students in grades 11 and 12 combined.
student_df.loc[(student_df["grade"] == 11) + (student_df["grade"] == 12) & (student_df["reading_score"])][["reading_score"]].mean()

reading_score    74.900381
dtype: float64

## Deliverable 5: Make Comparisons Between District and Charter Schools

Compare district vs charter schools for budget, size, and scores.

Make comparisons within your data by completing the following steps:

1. Using the `groupby` and `mean` functions, look at the average reading and math scores per school type.

1. Using the `groupby` and `count` functions, find the total number of students at each school.

2. Using the `groupby` and `mean` functions, find the average budget per grade for each school type.

In [22]:
# Use groupby and mean to find the average reading and math scores for each school type.
avg_student_scores_by_grade = student_df.groupby(by='school_type').mean()
avg_student_scores_by_grade.loc[:, ["math_score", "reading_score"]]
  

math_score  reading_score
school_type                           
Charter       66.761883      72.450603
Public        62.951576      72.281219

In [23]:
# Use the `groupby`, `count`, and `sort_values` functions to find the
# total number of students at each school and sort from most students to least students.

school_student_count = student_df.groupby(by= "school_name").count().sort_values(["student_name"], ascending=False)
school_student_count.loc[:, "student_name"]


school_name
Montgomery High School    2038
Green High School         1961
Dixon High School         1583
Wagner High School        1541
Silva High School         1109
Woods High School         1052
Sullivan High School       971
Turner High School         846
Bowers High School         803
Fisher High School         798
Richard High School        551
Campos High School         541
Odonnell High School       459
Campbell High School       407
Chang High School          171
Name: student_name, dtype: int64

In [24]:
avg_budget = student_df.groupby(by='school_type').mean()
avg_budget.loc[:, ["school_budget"]]

school_budget
school_type               
Charter      872625.656236
Public       911195.558251

In [25]:
avg_grade = student_df.groupby(by='school_type').mean()
avg_grade.loc[:, ["grade"]]

grade
school_type           
Charter      10.346744
Public       10.362808

In [26]:
#Observation of school type's math and reading scores vs budget
avg_student_scores_by_grade = student_df.groupby(by='school_type').mean()
avg_student_scores_by_grade.loc[:, ["math_score", "reading_score", "school_budget"]]

math_score  reading_score  school_budget
school_type                                          
Charter       66.761883      72.450603  872625.656236
Public        62.951576      72.281219  911195.558251

In [27]:
#How can I analyze school type, grade level, scores, and budget?


# Deliverable 6: Summarize Your Findings
In the cell below, write a few sentences to describe any discoveries you made while performing your analysis along with any additional analysis you believe would be worthwhile.

    The "Student Data Challenger Starter Code" data was analyzed with the use of Pandas and exercised in a Jupyter notebook. Upon analysis of the "Student Data Challenger Starter Code" the average math and reading scores were similar for both types of schools. The Public and Charter schools had an average reading score of 72%. Albeit, close in results the average math scores did slightly differ. Mathematical scores fo Public schools came in at 62% and Charter schools came in at 66%. When considering budget and it's causual relationship with grade averages it does denote some concern, considering the average budget for the charter schools is $38,570 on average. This could beg one to consider class room sizes, teacher salaries, teaching methods, types classrom materials, etc as possible variables that could be causing Charter schools to slightly out perform Public schools (on average) when observing the differences in budgets. 
    
    Overall, this was an important challenge to work through. Initially, the importing of data was not only important, but verifying that it was properly imported was equally crucial. This was handled through the .head() function which allowed me to analyze the data set in a data frame format with the necessary columns and first five rows of the data set. From this point, it was necessary to verify and remove missing values (.insa().sum() & .dropna()). Such functions allows one to quickly clean up the data to produce more reliable results. According to Data Independent (N.D.), "It is super helpful when you want to make sure you data has a unique key or unique rows." In addition to this procedure, I checked and removed duplicate rows (.duplicated().sum() & .drop_duplicates(inplace=True)) and verified data types with .dtypes to identify integers, floating numbers, and objects. Such clean up in data also may require removing certain entries within the data to allow the analyst to disiminate data even further. At this point in my research it was important to remove non-numeric characters from the school grades, because initially the school grade column contained objects due to the "th" following the numerical value of the grade. The letters were then removed later by using .str.replace('th', '') allowing only a numerical value being present in the student grade column, thus changing student grade from object to int64. 
   Midway through the analysis it was vital to then summarize the data through the .describe(), .mean(), and .min() functions. According to Melissa Rodrigues (2020), "describe(): generates descriptive statistics that will provide visibility of the dispersion and shape of a dataset’s distribution." For reporting reasons, the emprical data underwent analysis by finding the average mathematical score  and the minimum reading score for the entire data set. These methods sparked interest in the data set that could embark on further questioning of the data set in a way that contributes to the research overall. For instance, based on the provided findings I have produced another individual may be able to mirror the practices to analyze this data set. They could do this by collecting real world school data and analyze mathematical and reading scores versus school fundings on only a school level or a grade and school level. However, for ethical reasons it is important to note that this research more than likely already exists and may not be able to be published unless one finds a phenomena within the data that has not been explored yet. If that is the case I encourage the individual(s) to take this further if the uncharted phenomena is found. If not for scholarly research, this is still a usefull data set for any one practicing Data Analytics, especially when they want to exercise inferring upon causual relationships. In addition, they could take it further and compare Public versus Charter schools under the same perameters as well. To do so, it is useful to use .iloc[] and .loc[] as I have throughout the script, because it allows an individual to structure the data frame in a way that is succinct and customizable. However, it is important to keep in mind ethics at this point and ensure that the data frames are not altered in such a way to fit a bias. According to Shoonenboom and Burke Johnson (2017), "to design a mixed study, researchers must understand and carefully consider each of the dimensions of mixed methods design, and always keep an eye on the issue of validity." Finally, going back to the initial thought of parcing through data using .loc[] and .iloc[]. I used .groupby() and .sort_values() to dissiminate the data even further because I wanted to drill down to finding mean scores for Charter and Public schools, the total count (.count()) of students at the schools,  and the average budget of the Charter and Public schools. When using .groupby() the analyst, can aggregate data with functions such as: sum, max, min, and many other functions as well, (Rodriguez, 2020). 
   In closing, the findings in this data set drew up causual relationships between Charter and Public schools' budgets and grades. This was found through proper data collection, summarizing, and data analytics parsing techniques through Pandas fuctions. As mentioned, such data could be contributed to the community through further research to infer upon educational performance between the two types of schools. This would possibly be best demonstrated in a mixed methods study, but could be done through quantitative analysis alone if one does not want to capture thoughts and feelings of the participants and only numerical values of the subjects. According to Schonenboom and Burke Johnson (2017) "mixed methods research (“Mixed Methods” or “MM”) is the sibling of multimethod research (“Methodenkombination”) in which either solely multiple qualitative approaches or solely multiple quantitative approaches are combined." However, in order to do this one may have to compile more research based on teaching styles and methods of the schools and the school types overall by usings the .join() function within Python. 
   
   

References:

Data Independent. (N.D.). Pandas Drop Duplicates - pd.df.drop_duplicates(). https://dataindependent.com/pandas/pandas-drop-duplicates-pd-df-drop_duplicates/

Rodriguez, M. (2020, July 28). How to Summarize Data with Pandas. Medium.com. Analytics Vidhya.https://medium.com/analytics-vidhya/how-to-summarize-data-with-pandas-2c9edffafbaf#dee7

Schoonenboom, J. & Burke Johnson, R. (2017, July 5). NCBI. How to Construct a Mixed methods Research Design.https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5602001/. 107-131 doi: 10.1007/s11577-017-0454-1
    